### DISCUSSION TITLE
Possibility of adding additional data to hyperformula cells

### DISCUSSION THREAD

**Discussion by MartinDawson** [handsontable/hyperformula#739] [Feature requests]
I think this request would require a data structure change and it probably isn't trivial to do that because all the docs would need updating but I thought I'd raise it anyway to see what you guys thought.

So currently we are only able to use setCellContents with something like this:

`hf.setSellContents(cellAddress, [['=A1*B1']])`

What would be useful would be to allow storing of additional data alongside it, i.e like this:

`hf.setSellContents(cellAddress, [
  {
    height: 35,
    cells:  [{ text: '=A1*B1', border: 'solid 1px blue', backgroundColor: 'red' }]
  }
])`

So essentially making it an array of objects where the user can add anything inside it but the `cells` key is required and the `text` key is required.

The reason I would like this is because each cell has the ability to be styled by the user. Currently I have to have this in a separate array of objects:

```js
  rows: [
      {
        height: 35,
        cells: [
          {
             border: 'solid 1px blue', 
             backgroundColor: 'red' 
          },  
        ],
      },
    }
  ]
```

**Why would I need this?**

One example is if the user does a copy/paste into different cells (like in excel), hyperformula handles the formula and text moving but it doesn't handle any styles moving from cell to cell (i.e additional data). This means I have to keep a copy of the style data in the same format like above but also make sure the styles are copied to the same cells that hyperformula has done.

A single source of truth would be better so that my styles don't get out of sync with my formulas so if a copy/paste happens and I put my styles inside hyperformula it will automatically move them to the correct cell.

Thanks

**Comment by wojciechczerniak** [comment]
Thanks @MartinDawson for the idea. I'm marking it with Feature label, and maybe in the future we will reconsider. For now, we see HF as purely calculation engine, so storing additional data is out of scope for this project. Presentational attributes are more like another abstraction layer over the engine. A good practice says that View should be separate from Model.

Keeping such data would have an unnecessary overhead. Engine is mostly build for tracking dependencies and resolving changes between them. Static data like this does not require this overhead and could quickly clog up the browser memory. So as long as they do not need dependency tracking they should be stored next to the calculation data. Otherwise we would have to create nodes for empty cells that have styling attributes.

Taking the above under consideration, metadata like that should be stored next to the data graph to avoid overhead and synced anyway. So it's better to keep it outside of HF.

**Reply by MartinDawson** [reply]
Thanks for explaining, sounds good.

**Comment by asananddevsingh** [comment]
@wojciechczerniak 

This is something I am also looking for, Is this feature available now?

**Reply by sequba** [reply]
@asananddevsingh, the HyperFormula team still agrees with @wojciechczerniak's view on this. To keep the engine architecture clean and performant, we do not want to keep additional cell metadata. It should be stored outside HyperFormula.
